src/lib.rs

```diff
@@ -501,6 +616,102 @@ where
     }
 }
 
+impl<T, const BITS: usize> Mul for UInt<T, BITS>
+where
+    Self: Number,
+    T: PartialEq
+        + Copy
+        + BitAnd<T, Output = T>
+        + Not<Output = T>
+        + Mul<T, Output = T>
+        + Sub<T, Output = T>
+        + Shr<usize, Output = T>
+        + Shl<usize, Output = T>
+        + From<u8>,
+{
+    type Output = UInt<T, BITS>;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        let product = self.value * rhs.value;
+        #[cfg(debug_assertions)]
+        if (product & !Self::MASK) != T::from(0) {
+            panic!("attempt to multiply with overflow");
+        }
+        Self {
+            value: product & Self::MASK,
+        }
+    }
+}
+
+impl<T, const BITS: usize> MulAssign for UInt<T, BITS>
+where
+    Self: Number,
+    T: PartialEq
+        + Eq
+        + Not<Output = T>
+        + Copy
+        + MulAssign<T>
+        + BitAnd<T, Output = T>
+        + BitAndAssign<T>
+        + Sub<T, Output = T>
+        + Shr<usize, Output = T>
+        + Shl<usize, Output = T>
+        + From<u8>,
+{
+    fn mul_assign(&mut self, rhs: Self) {
+        self.value *= rhs.value;
+        #[cfg(debug_assertions)]
+        if (self.value & !Self::MASK) != T::from(0) {
+            panic!("attempt to multiply with overflow");
+        }
+        self.value &= Self::MASK;
+    }
+}
+
+impl<T, const BITS: usize> Div for UInt<T, BITS>
+where
+    Self: Number,
+    T: PartialEq
+        + Copy
+        + BitAnd<T, Output = T>
+        + Not<Output = T>
+        + Div<T, Output = T>
+        + Sub<T, Output = T>
+        + Shr<usize, Output = T>
+        + Shl<usize, Output = T>
+        + From<u8>,
+{
+    type Output = UInt<T, BITS>;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        // Integer division can only make the value smaller. And as the result is same type as
+        // Self, there's no need to range-check or mask
+        Self {
+            value: self.value / rhs.value,
```

Comment:
Does the compiler know this though? Should we use an unchecked version to guarantee we avoid checks?

Reply:
There's unfortunately not a Trait for wrapping_div in regular Rust, so we can't express the generic in this way (num_traits has that, but that's an optional feature in this crate)

Reply:
I don't think it matters too much though as I don't think the compiler checks anything here (what would it even test?)